[feat] Float comment window

joassouza opened the pull request:
This is the first implementation of the feature that allows showing the comment info dialog **(CIG)** on the Editor.
When user clicks in a comment icon and the support column is minimized, it shows the CIG on the Editor body.

### New behaviors 

- [x] edit comment on CIG

- [x] remove comment on CIG

- [x] listen to event from Teksto that informs when it should display the CIG

### Future Work

- [x] display more info on CIG (author, date, etc)

- [x] show length of replies

- [x] show replies text

- [x] add replies

- [x] reuse the code that controls the CIG on ep_touches

- [ ] improvements on UI/UX

Requires https://github.com/storytouch/web/tree/feat/text-marks-display-handler to work.

For more details [trello](https://trello.com/c/84i3zRkJ/1056-p3-coment%C3%A1rios-aplicar-melhorias-das-janelas-flutuantes-j%C3%A1-implementada-nas-tags-para-os-coment%C3%A1rios)

joassouza committed:
[feat] Show comment window when receive message via API

In this commit we listen to an event that makes the comment window to be
showed on the script body. We have a really basic implementation of the
comment data. Only the comment description is displayed.

joassouza committed:
[feat] Edit comment via float comment window

In this commit we did some refactor of the existent code. To make easier
we change the api.js to a prototype-like object. Some changes were made
on index.js as well. All of them to make easier to access the objects
and avoid conflict between objects

joassouza committed:
[feat] Delete comment via float window

joassouza committed:
[refactor] Rename event constant name

joassouza committed:
[refactor] Minor changes

joassouza committed:
[feat] Add translations to pop up buttons

joassouza committed:
[refactor] Minor changes

joassouza commented:
@lpagliari vou usar essa branch como base para as próximas branches.

lpagliari commented on static/js/commentIcons.js:
```diff
@@ -158,7 +158,7 @@ commentIcons.prototype._addIcon = function(commentId) {
 // Update which comments have reply
 commentIcons.prototype._updateCommentIconsStyle = function() {
-  var commentDataManager = pad.plugins.ep_comments_page.commentHandler.commentDataManager;
+  var commentDataManager = pad.plugins.ep_comments_page.commentDataManager;
```
dá para usar o `thisPlugin` aqui tb

lpagliari commented on static/js/copyPasteEvents.js:
```diff
@@ -35,7 +34,6 @@ exports.init = function() {
   pad.plugins = pad.plugins || {};
   pad.plugins.ep_comments_page = pad.plugins.ep_comments_page || {};
   pad.plugins.ep_comments_page.copyPasteHelpers = pad.plugins.ep_comments_page.copyPasteHelpers || [];
-  pad.plugins.ep_comments_page.fakeIdsMapper = fakeIdsMapper.init();
```
o `fakeIdsMapper` só é usado para o copy-paste... A idéia aqui era deixar todas as inicializações em um lugar só, é isso?

joassouza replied:
o fakeId é usado dentro do `shared.js` também. Eu gostei bastante do padrão que você adotou no ep_touches. Tudo centralizado, somente um lugar para controlar quem inicia primeiro

lpagliari commented on static/js/index.js:
```diff
@@ -488,7 +510,7 @@ exports.aceInitialized = function(hook, context){
 
 exports.aceRegisterNonScrollableEditEvents = function(){
-  return [preTextMarker.MARK_TEXT_EVENT, preTextMarker.UNMARK_TEXT_EVENT];
+  return [preTextMarker.MARK_TEXT_EVENT, preTextMarker.UNMARK_TEXT_EVENT, textMarkInfoDialog.NON_SCROLLABLE_EVENT];
```
melhor chamar o `NON_SCROLLABLE_EVENT` de um nome que explique mais o que ele faz, como nos outros eventos que definimos que não são "scrolláveis". Assim dá para entender o motivo desse cara existir e estar sendo passado aqui, eu acho...

Pensei em algo como `SELECT_TEXT_OF_COMMENT`, ou algo nessa linha... Veja o que vc acha que fica melhor.

lpagliari replied:
Vi aqui que o evento já tinha esse nome no ep_tags. Se quiser deixar com esse nome por enquanto, para facilitar enquanto os arquivos estão duplicados, não vejo problema. Podemos renomear a variável depois, quando o código estiver mais estável e limpo.

joassouza replied:
ótimo! Acho que fica mais fácil deixando igual para perceber onde devemos mudar depois. A ideia é ter o mesmo nome para `comments e tags`.

lpagliari commented on static/js/utils.js:
```diff
@@ -123,3 +123,12 @@ var selectFullTextOfRepArray = function(repArr, ace) {
 exports.selectFullTextOfRepArray = selectFullTextOfRepArray;
+
+// TODO: we have the same function on ep_touches. We should remove there
+// and use this one
```
boa!

lpagliari commented on templates/comments.html:
```diff
@@ -23,3 +23,35 @@
+    <div id="text-mark-info">
+      <div class="comment-description">
+        <div class="comment-description-body">${description}</div>
+      </div>
```
Uma dúvida, que talvez o resto do código responda depois (ainda não vi tudo): vc não está mostrando o autor do comentário na modal? Acho que precisaria...

joassouza replied:
Sim, precisa. Além disso talvez data também. Eu deixei essa adição mais pra frente

lpagliari commented on static/tests/frontend/specs/api-showCommentInfo.js:
```diff
@@ -0,0 +1,24 @@
+  it('shows the dialog with comment info', function(done) {
+    helper
+      .waitFor(function() {
+        return helper.padOuter$('#text-mark-info').is(':visible');
```
seria bom tb verificar se as informações do comentário estão sendo mostradas corretamente na modal

joassouza replied:
Sim, tem um conjunto de teste no próximo PR que checa essas informações adicionais (descrição). Deixei nesse teste só a validação da API.

joassouza replied:
Talvez fique um teste meio repetido... 🤔

lpagliari commented on static/tests/frontend/specs/workflowToEditComment.js:
```diff
@@ -0,0 +1,119 @@
+
+    it('opens the dialog to edit Comment', function(done) {
+      var test = this;
+      openCommentEditDialog(test, done);
```
como esse é um método para testar algo, melhor chamar de `testIfCommentEditDialogIsOpened`. Senão, chamando de `openCommentEditDialog`, fica parecendo que vc vai de fato *abrir* o CED

joassouza replied:
boa!

lpagliari commented on static/tests/frontend/specs/workflowToEditComment.js:
```diff
@@ -0,0 +1,119 @@
+    });
+
+    it('closes the Comment info dialog', function(done) {
+      closeCommentInfoDialog(done);
```
idem aqui, melhor renomear para `test...`

lpagliari commented on static/tests/frontend/specs/workflowToEditComment.js:
```diff
@@ -0,0 +1,119 @@
+        closeCommentEditDialog(done);
+      });
+
+      it('opens the Comment info dialog with the updated description', function(done) {
```
boa!!

lpagliari commented on static/tests/frontend/specs/workflowToEditComment.js:
```diff
@@ -0,0 +1,119 @@
+    });
+  });
+
+  // assume dialogs are opened
```
bom comentário! E bons métodos auxiliares tb, deixou o código bem mais limpo

lpagliari commented on static/tests/frontend/specs/workflowToRemoveComment.js:
```diff
@@ -0,0 +1,54 @@
+
+  var getCommentInfoDialog = function() {
+    return helper.padOuter$('.ui-dialog--comment:has(#text-mark-info)');
+  };
```
ambos os métodos foram definidos no outro teste tb. Melhor movê-los para o `utils` dos testes.

lpagliari commented on static/js/textMarkInfoDialog.js:
```diff
@@ -0,0 +1,217 @@
+
+textMarkInfoDialog.prototype.getCurrentOwner = function() {
+  return this.currentOwner;
+};
```
aaaaaaacho que esse cara não é mais usado (fiz uma busca aqui e não achei nada). Eu colocaria um `TODO` aqui para revisar isso depois

joassouza replied:
boa!

lpagliari commented on static/js/textMarkInfoDialog.js:
```diff
@@ -0,0 +1,217 @@
+  removeTextMark: function that deletes the text mark
+*/
+
+var textMarkInfoDialog = function(props) {
```
cara, ficou MUITO bom esse módulo!!! 👏 Sei que vc copiou uma parte do ep_tags, mas mesmo assim foi muito bacana a maneira que vc deixou partes customizáveis. Ótimo código!!

lpagliari commented on static/js/commentInfoDialog.js:
```diff
@@ -0,0 +1,54 @@
+    saveTextMark: this._saveComment.bind(this),
+    removeTextMark: this._removeComment.bind(this),
+  });
+  this.showCommentInfoForId = this.showCommentInfoForId.bind(this);
```
hum, esse `bind` ficou meio estranho aqui, não fica claro pq ele é necessário dentro desse módulo... 

Imagino que vc tenha precisado dele pq centralizou as definições dos callbacks da API no `index.js`, certo? Nesse caso, será que não faria mais sentido fazer o `bind` lá?

joassouza replied:
é desnecessário esse `bind` aí. Eu removi 👍

lpagliari reviewed (changes requested):
Cara, **MUITO** bom esse PR!! Vc conseguiu montar uma base bem limpa e clara (me pareceu) sobre as mudanças, acho que vai facilitar muito os próximos PRs dessa funcionalidade.

Sei que a parte do info dialog é meio complexa pq tem muita customização passando de um lado para o outro (eu sempre me perco tentando entender onde vai parar cada parte), mas assumindo que as modais estão configuradas direitinho, fica fácil construir novas funcionalidades em cima disso.

Muito bom PR! Ótima forma de terminar a semana :-)

joassouza committed:
[feat] Show reply window

In this commit we add a button with the length of replies as part of its
title. When it's clicked we render the window with the replies. For now,
we only display the reply text. In the next commits we'll add more
information and improve the UI.

joassouza committed:
[refactor] Address code review suggestions

joassouza committed:
[feat] Implement remove reply

In this commit we add the feature of removing a comment reply

joassouza committed:
[feat] Edit reply

In this commit we handle the edition of replies. We need to treat edges
cases as trying to save an empty text.

joassouza committed:
[feat] Does not allow empty text

joassouza committed:
[feat] Implement cancel edition of reply

joassouza committed:
[feat] Add more fields to comment window

In this commit we add the author, initials, scene where the comment is
to the comment info dialog

joassouza committed:
[feat] Add date field on comment info dialog

joassouza committed:
[feat] Add additional fields to reply info window

joassouza committed:
[feat] Toggle show/hide replies button text

joassouza committed:
[feat] Hide show replies button when there is not replies

joassouza committed:
[test] Improve tests readability

joassouza committed:
[refactor] Minor changes

In this commit we fix the translations and update the time of edition of
a reply on the client side.

joassouza committed:
[refactor] Address code review suggestions

joassouza committed:
[feat] Implement cancel reply addition

joassouza committed:
[feat] Save reply when comment has no replies

With this commit we handle other scenarios as well but the expected UI
state is broken

joassouza committed:
[feat] Control visibility of reply container after addition

In this commit we implement the UX of the reply addition. When the
comments already has replies and they're visible we keep the reply
dialog visible. When user adds the reply for the first time we make the
reply dialog visible as well.

joassouza committed:
[refactor] Minor changes

joassouza committed:
[style] Apply design on float comment window (#43)

* [design] add styles to comment and replies on the float window

* [refactor] Fix broken tests

This commit fixes the test that got broken by the previous commit

joassouza committed:
[refactor] Fix broken tests

joassouza committed:
[fix] Fix bug of show/hide replies button text

In this commit we fix the error that was displaying always the button
text as "hide replies". This bug was caused because the introduction of
animation. The animation has created a step between show and hide. So, for
example, in the scenario of changing from visible to hidden
the window animation would take some seconds to change its visibility.
As we query this visibility to decide if the button should have "show
replies or hide replies" as its text, we were not getting the final state
of the window

Roger-Siqueira committed:
[design] add styles to replying area of float window

joassouza committed:
[refactor] Address code review suggestions

joassouza committed:
[fix] Update comment UI after removal

joassouza committed:
[refactor] Address code review suggestions

joassouza committed:
[feat] Place focus at the end of text on edition

In this commit we place the focus on decription field when user edits a
text mark. The same feature is shared with comments and tags. We only
moved the same code from ep_touches to this plugin, so we can use in
both of them

joassouza committed:
[fix] Remove all the forms when recreate dialog

This fix the error of showing more than one replies form per comment
dialog

joassouza committed:
[refactor] Address code review suggestions

joassouza committed:
[refactor] Address code review suggestions

We don't need an unique id for reply field since we only can remove one
field per comment dialog

joassouza committed:
[feat] Close comment info dialog when click outside

It closes the comment info dialog when user clicks outside of the
comment info dialog.

This is part of https://trello.com/c/84i3zRkJ/1056

joassouza committed:
[feat] Avoid lose comment selection when click inside dialog

lpagliari reviewed (approved):
LGTM
